docs/dev/tombstone.md

```diff
@@ -0,0 +1,572 @@
+# Tombstones
+
+This document explains what tombstones are, how they used and how they are are created in ScyllaDB.
+
+## Introduction
+
+In an LSM Tree data structure, deleting data directly is not possible.
+Instead, deleting a key is translated to a write of a special deletion marker for said key.
+When the key and its deletion marker are compacted together, the key and any data it may have is dropped as a result.
+
+In ScyllaDB these deletion markers are called "tombstones".
+
+A tombstone has the following fields:
+
+```c++
+struct tombstone {
+    api::timestamp_type timestamp;
+    gc_clock::time_point deletion_time;
+}
+```
+
+The `timestamp` field is used to determine whether the tombstone covers some live data.
+If `tombstone.timestamp >= live_data.timestamp`, the data is covered by the tombstone and it is considered dead.
+In ScyllaDB, every cell has its own timestamp, so this coverage check happens on a cell level.
+It is possible that in a given row, some cells are covered, while others are not and they remain live.
+
+The `deletion_time` is used to determine whether the tombstone is eligible for garbage collection.
+Once a tombstone was compacted together with *all* data it could possibly cover and therefore all such data was dropped, it is no longer useful and can be garbage collected (dropped).
+Tombstone garbage collection involves complex rules, for more details see the [Tombstone Garbage Collection](./tombstone.md#tombstone-garbage-collection) chapter.
+
+## Tombstone Hierarchy
+
+Tombstones and data form a hierarchical structure, where tombstones can cover data on the same level and below, but not above.
+The tombstone hierarchy follows the data hierarchy, which looks like this:
+```mermaid
+graph TD;
+    partition-->row;
+    row-->cell;
+    row-->collection;
+    collection-->cell;
+```
+
+Tombstone hierarchy:
+```mermaid
+graph TD;
+    P[partition tombstone]-->RT[range tombstone];
+    RT[range tombstone]-->R[row tombstone];
+    R[row tombstone]-->C[cell tombstone];
+    R-->CT[collection tombstone];
+    CT-->C;
+```
+
+Partition tombstones applies to everything in the partition, including the static row.
+
+The range tombstone applies to a clustering range, so it can apply to multiple rows.
+
+Collections have a collection tombstone which applies to all elements in the collection.
+Collections members are similar to regular cells, they can be live or dead (cell tombstone).
+In the ScyllaDB type system, in addition to the `list`, `set` and `map` collection types, `tuple` and `UDT` are also stored as collections so the same rules apply to them as well.
+
+When determining whether some entity is live or not, one has to consider all tombstones that are above this entity in the tombstone hierarchy.
+For example, to determine whether a regular cell is live or not, one has to calculate the following tombstone: `partition tombstone + range tombstone + row tombstone`.
```

Comment:
+ cell tombstone?

Reply:
If the cell is dead already, no point in applying other tombstones on it, it won't make a difference.